### PR TITLE
Revert ANF changes and add a testcase for evaluation order

### PR DESCRIPTION
**Pull request by cocreature**
After careful consideration, we decided that the change in evaluation
order that was accidentally introduced by the ANF changes should be
considered a breaking change or arguably even a bug and should not
land in 1.3.0.

Therefore, this PR reverts the following commits:

1. 353d0da6f705ed64d8406ad6907dad355feac975
2. a45b51042f44a90f42b0e1c1a99beefae9f100a9
3. 04c7b2af7f7e07605d08f7f98e4ce98fe5dfeaae
4. a624dd72426f0eb1db407462d6cf7f7d904ca514
5. b3aab72cee4749b5221f4b0c7bef478986c212e6

Other PRs mostly had trivial merge conflicts that I resolved. The two
most interesting ones here are probably

1. https://github.com/digital-asset/daml/pull/6576 which was easy to
   resolve and the change to return SEValue instead of SExpr is still
   nice and useful even if we do not need the guarantees.
2. it https://github.com/digital-asset/daml/pull/6542 which required
   some changes since the constructors changed. If you want to review
   those changes in detail (they are pretty straightforward so not too
   important), it’s probably easiest to check out this PR and run
   ```
   git diff 2cd2a8f2a8052e9caa83dfd1fa36b2273b77b04e
   daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala
   ```
   to see the diff to the parent commit of the first commit that
   introduced ANF.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/tests/daml-test-files/EvaluationOrder.daml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+-- @ERROR Aborted: BOOM
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/EvaluationOrder.daml:1]: Don't we need a copyright header here?

Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/EvaluationOrder.daml:1]: :confused: CI doesn’t seem to think so?

Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/EvaluationOrder.daml:4]: Added one anyway :shrug:

Comment (garyverhaegen-da) [reply compiler/damlc/tests/daml-test-files/EvaluationOrder.daml:4]: There's a `NO_AUTO_COPYRIGHT` file in that folder. Given the number of files in there I suppose we should try to make that more granular at some point.

**Comment by garyverhaegen-da** [comment]
Assuming we consider this a temporary measure until we evaluate further, rather than a final decision to never do ANF, would it be possible to do the revert by hiding the ANF transform behind a (hardcoded) feature flag rather than delete everything? It would make further work on evaluating it and possibly trying to fix the evaluation order much easier.

File: compiler/damlc/tests/daml-test-files/EvaluationOrder.daml (hunk 1 above)
Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/EvaluationOrder.daml:4]: It doesn’t matter much. They are all just test files so I don’t think having the copyright header there is particularly important. I would treat them similar to our templates.

**Comment by cocreature** [comment]
> would it be possible to do the revert by hiding the ANF transform behind a (hardcoded) feature flag rather than delete everything

Not really, it’s a pretty invasive change in speedy and speedy just bails out if your code isn’t in ANF. Making that work with a feature flag is a lot of work especially if you want to keep the performance benefits and at that point I think the efforts are better spent on fixing the issue and reintroducing the transformation.

**Pull request merged**